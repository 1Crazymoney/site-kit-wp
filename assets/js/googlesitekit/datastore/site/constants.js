export const STORE_NAME = 'core/site';
