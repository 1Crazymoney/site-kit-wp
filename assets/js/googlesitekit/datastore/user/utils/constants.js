export const INVALID_DATE_INSTANCE_ERROR = 'Date param must construct to a valid date instance or be a valid date instance itself.';
export const INVALID_DATE_STRING_ERROR = 'Invalid dateString parameter, it must be a string.';
