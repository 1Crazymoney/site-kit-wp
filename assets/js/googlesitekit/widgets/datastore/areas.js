/**
 * core/widgets data store: widgets info.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';

/**
 * Internal dependencies
 */
import {
	WIDGET_STYLES,
	CONTEXT_DASHBOARD,
	CONTEXT_PAGE_DASHBOARD,
} from './constants';

/**
 * Store our widget components by registry, then by widget `slug`. We do this because
 * we can't store React components in our data store.
 *
 * @private
 * @since 1.9.0
 */
export const WidgetComponents = {};

const ASSIGN_WIDGET_AREA = 'ASSIGN_WIDGET_AREA';
const REGISTER_WIDGET_AREA = 'REGISTER_WIDGET_AREA';

const WidgetStyleKeys = Object.keys( WIDGET_STYLES ).map( ( ( key ) => `WIDGET_STYLES.${ key }` ) ).join( ', ' );

export const INITIAL_STATE = {
	areas: {},
	contextAssignments: {},
};

export const actions = {
	/**
	 * Assigns a widget area to one (or several) contexts.
	 *
	 * Accepts an area slug to register as the first argument, then either a string
	 * (for a single context slug) or array of contexts slugs (to assign the widget area
	 * to multiple contexts).
	 *
	 * Does not error if any of the requested contexts do not exist, but will console.warn
	 * a message about non-existant contexts.
	 *
	 * Does not error if the area slug provided does not exist, but will console.warn a
	 * message about non-existent areas.
	 *
	 * @since 1.9.0
	 *
	 * @param {string}         slug         Widget Area's slug.
	 * @param {(string|Array)} contextSlugs Widget Context's slug(s).
	 * @return {Object} Redux-style action.
	 */
	assignWidgetArea( slug, contextSlugs ) {
		return {
			payload: {
				slug,
				contextSlugs: ( typeof contextSlugs === 'string' ) ? [ contextSlugs ] : contextSlugs,
			},
			type: ASSIGN_WIDGET_AREA,
		};
	},

	/**
	 * Creates a widget area with a given name and settings.
	 *
	 * @since 1.9.0
	 *
	 * @param {string}             slug               Widget Area's slug.
	 * @param {Object}             settings           Widget Area's settings.
	 * @param {string}             settings.title     Title for this widget area.
	 * @param {string}             settings.subtitle  Subtitle for this widget area.
	 * @param {string}             settings.icon      Optional. URL to SVG icon for this widget area.
	 * @param {string}             settings.style     Optional. Widget area style (one of "boxes", "composite"). Default: "boxes".
	 * @param {number}             settings.priority  Optional. Priority for this widget area. Default: 10.
	 * @return {Object} Redux-style action.
	 */
	registerWidgetArea( slug, {
		priority = 10,
		style = WIDGET_STYLES.BOXES,
		title,
		subtitle,
		icon,
	} = {} ) {
		invariant( slug, 'slug is required.' );
		invariant( title, 'settings.title is required.' );
		invariant( subtitle, 'settings.subtitle is required.' );
		invariant( Object.values( WIDGET_STYLES ).includes( style ), `settings.style must be one of: ${ WidgetStyleKeys }.` );

		return {
			payload: {
				slug,
				settings: { priority, style, title, subtitle, icon },
			},
			type: REGISTER_WIDGET_AREA,
		};
	},
};

export const controls = {};

export const reducer = ( state, { type, payload } ) => {
	switch ( type ) {
		case ASSIGN_WIDGET_AREA: {
			const { slug, contextSlugs } = payload;

			const { contextAssignments } = state;
			contextSlugs.forEach( ( contextSlug ) => {
				if ( contextAssignments[ contextSlug ] === undefined ) {
					contextAssignments[ contextSlug ] = [];
				}

				if ( ! contextAssignments[ contextSlug ].includes( slug ) ) {
					contextAssignments[ contextSlug ].push( slug );
				}
			} );

			return {
				...state,
				contextAssignments,
			};
		}

		case REGISTER_WIDGET_AREA: {
			const { slug, settings } = payload;

			if ( state.areas[ slug ] !== undefined ) {
				global.console.warn( `Could not register widget area with slug "${ slug }". Widget area "${ slug }" is already registered.` );

				return { ...state };
			}

			return {
				...state,
				areas: {
					...state.areas,
					[ slug ]: { ...settings, slug },
				},
			};
		}

		default: {
			return { ...state };
		}
	}
};

export const resolvers = {
	/**
	 * Defines default widget areas for a given context.
	 *
	 * @since n.e.x.t
	 *
	 * @param {string} contextSlug Widget context to get areas for.
	 */
	*getWidgetAreas( contextSlug ) {
		if ( contextSlug === CONTEXT_DASHBOARD ) {
			yield actions.registerWidgetArea( 'dashboardAllTraffic', {
				title: 'All Triffic',
				subtitle: 'How people found your site.',
			} );

			yield actions.registerWidgetArea( 'dashboardSearchFunnel', {
				title: 'Search Funnel',
				subtitle: 'How your site appeared in Search results and how many visitors you got from Search.',
			} );

			yield actions.registerWidgetArea( 'dashboardPopularity', {
				title: 'Popularity',
				subtitle: 'Your most popular pages and how people found them from Search.',
			} );

			yield actions.registerWidgetArea( 'dashboardSpeed', {
				title: 'Page Speed and Experience',
				subtitle: 'How fast your home page loads, how quickly people can interact with your content, and how stable your content is.',
			} );

			yield actions.registerWidgetArea( 'dashboardEarnings', {
				title: 'Earnings',
				subtitle: 'How much your site earns.',
			} );

			yield actions.assignWidgetArea( 'dashboardAllTraffic', CONTEXT_DASHBOARD );
			yield actions.assignWidgetArea( 'dashboardSearchFunnel', CONTEXT_DASHBOARD );
			yield actions.assignWidgetArea( 'dashboardPopularity', CONTEXT_DASHBOARD );
			yield actions.assignWidgetArea( 'dashboardSpeed', CONTEXT_DASHBOARD );
			yield actions.assignWidgetArea( 'dashboardEarnings', CONTEXT_DASHBOARD );
		} else if ( contextSlug === CONTEXT_PAGE_DASHBOARD ) {
			yield actions.registerWidgetArea( 'pageDashboardAllTraffic', {
				title: 'All Triffic',
				subtitle: 'How people found your site.',
			} );

			yield actions.registerWidgetArea( 'pageDashboardSearchFunnel', {
				title: 'Search Funnel',
				subtitle: 'How your site appeared in Search results and how many visitors you got from Search.',
			} );

			yield actions.registerWidgetArea( 'pageDashboardPopularity', {
				title: 'Popularity',
				subtitle: 'Your most popular pages and how people found them from Search.',
			} );

			yield actions.registerWidgetArea( 'pageDashboardSpeed', {
				title: 'Page Speed and Experience',
				subtitle: 'How fast your home page loads, how quickly people can interact with your content, and how stable your content is.',
			} );

			yield actions.registerWidgetArea( 'pageDashboardEarnings', {
				title: 'Earnings',
				subtitle: 'How much your site earns.',
			} );

			yield actions.assignWidgetArea( 'pageDashboardAllTraffic', CONTEXT_PAGE_DASHBOARD );
			yield actions.assignWidgetArea( 'pageDashboardSearchFunnel', CONTEXT_PAGE_DASHBOARD );
			yield actions.assignWidgetArea( 'pageDashboardPopularity', CONTEXT_PAGE_DASHBOARD );
			yield actions.assignWidgetArea( 'pageDashboardSpeed', CONTEXT_PAGE_DASHBOARD );
			yield actions.assignWidgetArea( 'pageDashboardEarnings', CONTEXT_PAGE_DASHBOARD );
		}
	},
};

export const selectors = {
	/**
	 * Checks if a widget area has been registered.
	 *
	 * Returns `true` if the widget area has been registered.
	 * Returns `false` if the widget area has NOT been registered.
	 *
	 * @since 1.9.0
	 *
	 * @param {Object} state Data store's state.
	 * @param {string} slug  Widget Area's slug.
	 * @return {boolean} `true`/`false` based on whether widget area has been registered.
	 */
	isWidgetAreaRegistered( state, slug ) {
		const { areas } = state;

		return areas[ slug ] !== undefined;
	},

	/**
	 * Returns all widget areas registered for a given context.
	 *
	 * Returns an array of all widget areas for a given context.
	 * The widget areas are returned in order of their priority, so can be rendered in
	 * the order provided by the selector.
	 *
	 * @since 1.9.0
	 *
	 * @param {Object} state       Data store's state.
	 * @param {string} contextSlug Widget context to get areas for.
	 * @return {Array} An ordered array of widget areas for this context.
	 */
	getWidgetAreas( state, contextSlug ) {
		invariant( contextSlug, 'contextSlug is required.' );

		const { areas, contextAssignments } = state;

		return Object.values( areas ).filter( ( area ) => {
			return contextAssignments[ contextSlug ] && contextAssignments[ contextSlug ].includes( area.slug );
		} ).sort( ( areaA, areaB ) => areaA.priority - areaB.priority );
	},

	/**
	 * Returns a widget area based on slug.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} state  Data store's state.
	 * @param {string} slug   Widget area to select.
	 * @return {Object|null} A widget area if one is found with a matching slug.
	 */
	getWidgetArea( state, slug ) {
		invariant( slug, 'slug is required.' );

		const { areas } = state;

		return areas[ slug ] || null;
	},
};

export default {
	INITIAL_STATE,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
};
