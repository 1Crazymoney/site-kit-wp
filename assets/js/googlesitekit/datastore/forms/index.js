/**
 * core/forms Data store: values
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import Data from 'googlesitekit-data';
import invariant from 'invariant';
import { STORE_NAME } from './constants';

export { STORE_NAME };

const SET_FORM_VALUES = 'SET_FORM_VALUES';

export const INITIAL_STATE = {};

export const actions = {
	setValues( formName, formData ) {
		invariant( formName, 'formName is required for setting values.' );

		return {
			payload: { formName, formData },
			type: SET_FORM_VALUES,
		};
	},
};

export const controls = {};

export const reducer = ( state, { type, payload } ) => {
	switch ( type ) {
		case SET_FORM_VALUES: {
			const { formName, formData } = payload;

			return {
				...state,
				[ formName ]: formData,
			};
		}

		default: {
			return { ...state };
		}
	}
};

export const resolvers = {};

export const selectors = {
	getValue( state, formName, key ) {
		return state[ formName ][ key ];
	},
};

const store = {
	INITIAL_STATE,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
};

// Register this store on the global registry.
Data.registerStore( STORE_NAME, store );

export default store;
