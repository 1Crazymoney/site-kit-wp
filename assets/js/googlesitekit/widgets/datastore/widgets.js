/**
 * core/widgets data store: widgets info.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { STORE_NAME } from './constants';

const { commonActions, createRegistrySelector } = Data;

/**
 * Store our widget components by registry, then by widget `slug`. We do this because
 * we can't store React components in our data store.
 *
 * @private
 * @since n.e.x.t
 */
export const WidgetComponents = {};

const ASSIGN_WIDGET = 'ASSIGN_WIDGET';
const REGISTER_WIDGET = 'REGISTER_WIDGET';
const SET_WIDGET_COMPONENT_KEY = 'SET_WIDGET_COMPONENT_KEY';

export const INITIAL_STATE = {
	registryKey: undefined,
	widgets: {},
};

export const actions = {
	/**
	 * Assigns an existing widget (by slug) to a widget area(s).
	 *
	 * @since n.e.x.t
	 *
	 * @param  {string}         slug      Widget slug.
	 * @param  {(string|Array)} areaSlugs Widget Area slug(s).
	 * @return {Object}                   Redux-style action.
	 */
	assignWidget( slug, areaSlugs ) {
		let areaSlugsAsArray;
		if ( typeof areaSlugs === 'string' ) {
			areaSlugsAsArray = [ areaSlugs ];
		} else {
			areaSlugsAsArray = [ ...areaSlugs ];
		}

		return {
			payload: { slug, areaSlugs: areaSlugsAsArray },
			type: ASSIGN_WIDGET,
		};
	},

	/**
	 * Register a widget with a given slug and settings.
	 *
	 * @since n.e.x.t
	 * @param {string} slug           Widget's slug.
	 * @param {Object} settings       Widget's settings.
	 * @return {Object}          Redux-style action.
	 */
	*registerWidget( slug, settings ) {
		const registry = yield commonActions.getRegistry();
		let registryKey = yield registry.select( STORE_NAME ).getWidgetRegistryKey();

		if ( registryKey === undefined ) {
			registryKey = Object.keys( WidgetComponents ).length + 1;
			yield {
				payload: { registryKey },
				type: SET_WIDGET_COMPONENT_KEY,
			};
		}

		// We do this assignment in the action rather than the reducer because we can't send a
		// payload that includes a React component to the reducer; we'll get an error about
		// payloads needing to be plain objects.
		if ( WidgetComponents[ registryKey ] === undefined ) {
			WidgetComponents[ registryKey ] = {};
		}
		if ( WidgetComponents[ registryKey ][ slug ] === undefined ) {
			WidgetComponents[ registryKey ][ slug ] = settings.component;
		}
		const settingsToUse = {
			priority: 10,
			...settings,
		};
		delete settingsToUse.component;

		yield {
			payload: { slug, settings: settingsToUse },
			type: REGISTER_WIDGET,
		};

		return {};
	},
};

export const controls = {};

export const reducer = ( state, { type, payload } ) => {
	switch ( type ) {
		case ASSIGN_WIDGET: {
			const { slug, areaSlugs } = payload;

			const { areaWidgets } = state;
			areaSlugs.forEach( ( areaSlug ) => {
				if ( areaWidgets[ areaSlug ] === undefined ) {
					areaWidgets[ areaSlug ] = [];
				}

				if ( ! areaWidgets[ areaSlug ].includes( slug ) ) {
					areaWidgets[ areaSlug ].push( slug );
				}
			} );

			return {
				...state,
				areaWidgets,
			};
		}

		case REGISTER_WIDGET: {
			const { slug, settings } = payload;

			if ( state.widgets[ slug ] !== undefined ) {
				global.console.warn( `Could not register widget with slug "${ slug }". Widget "${ slug }" is already registered.` );

				return { ...state };
			}

			return {
				...state,
				widgets: {
					...state.widgets,
					[ slug ]: { ...settings, slug },
				},
			};
		}

		case SET_WIDGET_COMPONENT_KEY: {
			const { registryKey } = payload;

			return {
				...state,
				registryKey,
			};
		}

		default: {
			return { ...state };
		}
	}
};

export const resolvers = {
};

export const selectors = {
	/**
	 * Checks if a widget has been registered with a given slug.
	 *
	 * Returns `true` if the widget area has been registered.
	 * Returns `false` if the widget area has NOT been registered.
	 *
	 * @since n.e.x.t
	 *
	 * @param  {Object}  state Data store's state.
	 * @param  {string}  slug  Widget's slug.
	 * @return {boolean}       `true`/`false` based on whether widget has been registered.
	 */
	isWidgetRegistered( state, slug ) {
		const { widgets } = state;

		return widgets[ slug ] !== undefined;
	},

	/**
	 * Returns all widgets registered for a given widget area.
	 *
	 * Returns an array of all widgets for a given area.
	 * The widgets are returned in order of their priority, so can be rendered in
	 * the order provided by the selector.
	 *
	 * @since n.e.x.t
	 *
	 * @param  {Object} state          Data store's state.
	 * @param  {string} widgetAreaSlug Widget context to get areas for.
	 * @return {Array}                 An ordered array of widgets for this area.
	 */
	getWidgets: createRegistrySelector( ( select ) => ( state, widgetAreaSlug ) => {
		invariant( widgetAreaSlug, 'widgetAreaSlug is required.' );

		const { areaWidgets, widgets } = state;

		const registryKey = select( STORE_NAME ).getWidgetRegistryKey();

		return Object.values( widgets ).filter( ( widget ) => {
			return areaWidgets[ widgetAreaSlug ] && areaWidgets[ widgetAreaSlug ].includes( widget.slug );
		} ).sort( ( widgetA, widgetB ) => {
			return ( widgetA.priority >= widgetB.priority ) ? 1 : -1;
		} ).map( ( widget ) => {
			const widgetWithComponent = { ...widget };
			if ( WidgetComponents[ registryKey ] ) {
				widgetWithComponent.component = WidgetComponents[ registryKey ][ widget.slug ];
			}

			return widgetWithComponent;
		} );
	} ),

	/**
	 * Returns the registry key being used for this registry's widgets.
	 *
	 * We key each registry with an Integer, so we don't share registered widgets
	 * between registries. This allows us to access the appropriate registry global
	 * from inside selectors.
	 *
	 * @private
	 * @since n.e.x.t
	 *
	 * @param  {Object}              state Data store's state.
	 * @return {(number|undefined)}        An ordered array of widgets for this area.
	 */
	getWidgetRegistryKey( state ) {
		const { registryKey } = state;

		return registryKey;
	},
};

export default {
	INITIAL_STATE,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
};
