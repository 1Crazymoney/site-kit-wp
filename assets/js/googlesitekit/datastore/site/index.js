/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
import { collect, collectReducers } from 'assets/js/googlesitekit/data/utils';
import connection from './connection';
import reset from './reset';

export const INITIAL_STATE = {
	connection: null,
	isFetchingConnection: false,
	isResetting: false,
};

export const STORE_NAME = 'core/site';

export const actions = collect(
	connection.actions,
	reset.actions,
);

export const controls = collect(
	connection.controls,
	reset.controls,
);

export const reducer = collectReducers( INITIAL_STATE, [
	connection.reducer,
	reset.reducer,
] );

export const resolvers = collect(
	connection.resolvers,
	reset.resolvers,
);

export const selectors = collect(
	connection.selectors,
	reset.selectors,
);

const store = {
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
};

export const registerOn = ( registry ) => {
	registry.registerStore( STORE_NAME, store );
};

export default store;
