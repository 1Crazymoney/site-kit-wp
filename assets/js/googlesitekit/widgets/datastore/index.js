/**
 * core/widgets data store
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import areas from './areas';
import widgets from './widgets';

export const INITIAL_STATE = Data.collectState(
	areas.INITIAL_STATE,
	widgets.INITIAL_STATE,
);

import { STORE_NAME } from './constants';
export { STORE_NAME };

export const actions = Data.addInitializeAction(
	Data.collectActions(
		Data.commonActions,
		areas.actions,
		widgets.actions,
	)
);

export const controls = Data.collectControls(
	Data.commonControls,
	areas.controls,
	widgets.controls,
);

export const reducer = Data.addInitializeReducer(
	INITIAL_STATE,
	Data.collectReducers(
		areas.reducer,
		widgets.reducer,
	)
);

export const resolvers = Data.collectResolvers(
	areas.resolvers,
	widgets.resolvers,
);

export const selectors = Data.collectSelectors(
	areas.selectors,
	widgets.selectors,
);

const store = {
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
};

// Register this store on the global registry.
Data.registerStore( STORE_NAME, store );

export default store;
