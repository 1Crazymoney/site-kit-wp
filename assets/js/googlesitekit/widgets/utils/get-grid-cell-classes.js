/**
 * Widgets: Grid Cell Layout utility functions.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { WIDTH_GRID_MAP, WIDGET_WIDTHS } from '../datastore/constants';

const resizeClasses = ( classNamesMap, counter ) => {
	[ ...classNamesMap ].reverse().some( ( _classNames, index ) => {
		const originalIndex = classNamesMap.length - 1 - index;

		if ( _classNames.includes( 'mdc-layout-grid__cell--span-3-desktop' ) ) {
			// Replace the 3-column class with a 4-column class so this element goes from 1/4
			// to 1/3 on desktop.
			classNamesMap[ originalIndex ][ _classNames.indexOf( 'mdc-layout-grid__cell--span-3-desktop' ) ] = 'mdc-layout-grid__cell--span-4-desktop';

			counter -= 3;
		}

		if ( _classNames.includes( 'mdc-layout-grid__cell--span-6-desktop' ) ) {
			// Replace the 6-column class with a 8-column class so this element goes from 1/2
			// to 2/3 on desktop.
			classNamesMap[ originalIndex ][ _classNames.indexOf( 'mdc-layout-grid__cell--span-6-desktop' ) ] = 'mdc-layout-grid__cell--span-8-desktop';

			counter -= 6;
		}

		return counter === 0;
	} );

	return [ classNamesMap, counter ];
};

/**
 * Returns an array of `classNames` to assign to a Widget wrapper.
 *
 * @since n.e.x.t
 *
 * @param {Array} widgets An array of `Widget` objects, usually from the `getWidgets()` selector.
 * @return {Array} Array of classNames to apply to an element.
 */
export const getGridCellClasses = ( widgets ) => {
	let counter = 0;
	let classNamesMap = [];

	widgets.forEach( ( widget, i ) => {
		const WidgetComponent = widget.component;
		const widgetOutput = <WidgetComponent />;

		// If this widget output `null`, there's no sense in outputting classes for it.
		if ( widgetOutput === null ) {
			return null;
		}

		const classNames = [ 'mdc-layout-grid__cell' ];

		if ( widget.width === WIDGET_WIDTHS.FULL ) {
			classNames.push(
				'mdc-layout-grid__cell--span-12',
			);
		}

		if ( widget.width === WIDGET_WIDTHS.HALF ) {
			classNames.push(
				'mdc-layout-grid__cell--span-6-desktop',
				'mdc-layout-grid__cell--span-8-tablet',
			);
		}

		if ( widget.width === WIDGET_WIDTHS.QUARTER ) {
			classNames.push(
				'mdc-layout-grid__cell--span-3-desktop',
				'mdc-layout-grid__cell--span-4-tablet',
			);
		}

		classNamesMap[ i ] = classNames;

		counter += WIDTH_GRID_MAP[ widget.width ];

		if ( counter % 12 === 0 ) {
			counter = 0;
		}

		if ( counter > 12 ) {
			counter -= WIDTH_GRID_MAP[ widget.width ];

			if ( counter === 9 ) {
				while ( counter !== 0 ) {
					[ classNamesMap, counter ] = resizeClasses( classNamesMap, counter );
				}
			}

			counter = WIDTH_GRID_MAP[ widget.width ];
		}

		if ( counter === 9 ) {
			while ( counter !== 0 ) {
				[ classNamesMap, counter ] = resizeClasses( classNamesMap, counter );
			}
		}
	} );

	return classNamesMap;
};
