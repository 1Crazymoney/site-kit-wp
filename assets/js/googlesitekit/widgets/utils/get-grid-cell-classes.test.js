/**
 * Widgets: Grid Cell Layout utility tests.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	unsubscribeFromAll,
	// muteConsole,
} from '../../../../../tests/js/utils';
// import { render } from '../../../../../tests/js/test-utils';
import { STORE_NAME, WIDGET_WIDTHS } from '../datastore/constants';
// import { WidgetComponents } from './widgets';
// import { WIDTH_GRID_MAP, WIDGET_WIDTHS } from '../datastore/constants';
import { getGridCellClasses } from './get-grid-cell-classes';

const WidgetComponent = () => {
	return ( <div>Foo bar!</div> );
};

const createWidgets = ( registry, widgets ) => {
	widgets.forEach( ( { component, slug, width } ) => {
		registry.dispatch( STORE_NAME ).registerWidget( slug, {
			component,
			width,
		} );
		registry.dispatch( STORE_NAME ).assignWidget( slug, 'gridcell-test' );
	} );
};

describe( 'getGridCellClasses', () => {
	let registry;
	// let store;

	beforeEach( () => {
		registry = createTestRegistry();

		registry.dispatch( STORE_NAME ).registerWidgetArea( 'gridcell-test', {
			title: 'Dashboard Header',
			subtitle: 'Cool stuff for yoursite.com',
			style: 'boxes',
		} );
		registry.dispatch( STORE_NAME ).assignWidgetArea( 'gridcell-test', 'dashboard' );
		// store = registry.stores[ STORE_NAME ].store;
	} );

	afterEach( () => {
		unsubscribeFromAll( registry );
	} );

	it( 'should return the same number of elements as widgets from a selector', () => {
		createWidgets( registry, [
			{ component: WidgetComponent, slug: 'one', width: WIDGET_WIDTHS.FULL },
			{ component: WidgetComponent, slug: 'two', width: WIDGET_WIDTHS.FULL },
			{ component: WidgetComponent, slug: 'three', width: WIDGET_WIDTHS.FULL },
		] );

		const widgets = registry.select( STORE_NAME ).getWidgets( 'gridcell-test' );

		expect( widgets ).toHaveLength( 3 );
		expect( getGridCellClasses( widgets ) ).toHaveLength( 3 );
	} );

	it( 'should return all full classNames when all widgets are full-width', () => {
		createWidgets( registry, [
			{ component: WidgetComponent, slug: 'one', width: WIDGET_WIDTHS.FULL },
			{ component: WidgetComponent, slug: 'two', width: WIDGET_WIDTHS.FULL },
			{ component: WidgetComponent, slug: 'three', width: WIDGET_WIDTHS.FULL },
		] );

		const widgets = registry.select( STORE_NAME ).getWidgets( 'gridcell-test' );

		expect( getGridCellClasses( widgets ) ).toMatchSnapshot();
	} );

	it( 'should return appropriate classNames for each width', () => {
		createWidgets( registry, [
			{ component: WidgetComponent, slug: 'one', width: WIDGET_WIDTHS.FULL },
			{ component: WidgetComponent, slug: 'two', width: WIDGET_WIDTHS.QUARTER },
			{ component: WidgetComponent, slug: 'three', width: WIDGET_WIDTHS.HALF },
		] );

		const widgets = registry.select( STORE_NAME ).getWidgets( 'gridcell-test' );

		expect( getGridCellClasses( widgets ) ).toMatchSnapshot();
	} );
} );
