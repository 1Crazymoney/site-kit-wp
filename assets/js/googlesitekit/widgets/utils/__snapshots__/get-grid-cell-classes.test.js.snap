// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGridCellClasses should return all full classNames when all widgets are full-width 1`] = `
Array [
  Array [
    "mdc-layout-grid__cell",
    "mdc-layout-grid__cell--span-12",
  ],
  Array [
    "mdc-layout-grid__cell",
    "mdc-layout-grid__cell--span-12",
  ],
  Array [
    "mdc-layout-grid__cell",
    "mdc-layout-grid__cell--span-12",
  ],
]
`;

exports[`getGridCellClasses should return appropriate classNames for each width 1`] = `
Array [
  Array [
    "mdc-layout-grid__cell",
    "mdc-layout-grid__cell--span-12",
  ],
  Array [
    "mdc-layout-grid__cell",
    "mdc-layout-grid__cell--span-4-desktop",
    "mdc-layout-grid__cell--span-4-tablet",
  ],
  Array [
    "mdc-layout-grid__cell",
    "mdc-layout-grid__cell--span-8-desktop",
    "mdc-layout-grid__cell--span-8-tablet",
  ],
]
`;
